File: 01-machine/02-threads-and-syscalls/golang/syscall_cost.go
// Layer 1 - What a syscall actually costs.
// Same read(/dev/zero) vs pure-loop comparison, calling the raw syscall
// package directly so there's no extra runtime wrapper in the way.
package main

import (
	"fmt"
	"os"
	"time"
)

const N = 500_000

func benchSyscall() time.Duration {
	f, err := os.Open("/dev/zero")
	if err != nil {
		panic(err)
	}
	defer f.Close()
	buf := make([]byte, 1)
	start := time.Now()
	for i := 0; i < N; i++ {
		f.Read(buf)
	}
	return time.Since(start)
}

func benchPureGo() (time.Duration, int) {
	total := 0
	start := time.Now()
	for i := 0; i < N; i++ {
		total += i & 0xFF
	}
	return time.Since(start), total
}

func main() {
	tSys := benchSyscall()
	tPure, _ := benchPureGo()
	fmt.Printf("N=%d\n", N)
	fmt.Printf("read(/dev/zero) x%d:  %6.3fs  (%6.1f ns/call)\n", N, tSys.Seconds(), float64(tSys.Nanoseconds())/N)
	fmt.Printf("pure go loop:         %6.3fs  (%6.1f ns/iter)\n", tPure.Seconds(), float64(tPure.Nanoseconds())/N)
	fmt.Printf("syscall is %.1fx the cost of an equivalent pure-go step\n", tSys.Seconds()/tPure.Seconds())
}

File: 01-machine/02-threads-and-syscalls/cpp/syscall_cost.cpp
// Layer 1 - What a syscall actually costs, C++ version.
// read(2) directly via POSIX <unistd.h> -- there is no wrapper at all
// between this code and the kernel, which makes it the most direct
// measurement of raw syscall cost in this whole lab.
#include <chrono>
#include <cstdio>
#include <fcntl.h>
#include <unistd.h>

constexpr long N = 500'000;

double bench_syscall() {
    int fd = open("/dev/zero", O_RDONLY);
    char buf[1];
    auto start = std::chrono::high_resolution_clock::now();
    for (long i = 0; i < N; i++) {
        read(fd, buf, 1);
    }
    auto end = std::chrono::high_resolution_clock::now();
    close(fd);
    return std::chrono::duration<double>(end - start).count();
}

double bench_pure_cpp(volatile long& sink) {
    long total = 0;
    auto start = std::chrono::high_resolution_clock::now();
    for (long i = 0; i < N; i++) {
        total += i & 0xFF;
    }
    auto end = std::chrono::high_resolution_clock::now();
    sink = total; // defeat dead-code elimination, same issue Rust's version hit
    return std::chrono::duration<double>(end - start).count();
}

int main() {
    volatile long sink = 0;
    double t_sys = bench_syscall();
    double t_pure = bench_pure_cpp(sink);
    std::printf("N=%ld\n", N);
    std::printf("read(/dev/zero) x%ld:  %6.3fs  (%6.1f ns/call)\n", N, t_sys, t_sys / N * 1e9);
    std::printf("pure C++ loop:         %6.3fs  (%6.1f ns/iter)\n", t_pure, t_pure / N * 1e9);
    std::printf("syscall is %.1fx the cost of an equivalent pure-C++ step\n", t_sys / t_pure);
    return 0;
}
